dubbo-common/src/main/java/org/apache/dubbo/common/threadpool/support/AbortPolicyWithReport.java

```diff
@@ -78,43 +74,34 @@ private void dumpJStack() {
             return;
         }
 
-        Executors.newSingleThreadExecutor().execute(new Runnable() {
-            @Override
-            public void run() {
-                String dumpPath = url.getParameter(Constants.DUMP_DIRECTORY, System.getProperty("user.home"));
-
-                SimpleDateFormat sdf;
-
-                String OS = System.getProperty("os.name").toLowerCase();
-
-                // window system don't support ":" in file name
-                if(OS.contains("win")){
-                    sdf = new SimpleDateFormat("yyyy-MM-dd_HH-mm-ss");
-                }else {
-                    sdf = new SimpleDateFormat("yyyy-MM-dd_HH:mm:ss");
-                }
-
-                String dateStr = sdf.format(new Date());
-                FileOutputStream jstackStream = null;
-                try {
-                    jstackStream = new FileOutputStream(new File(dumpPath, "Dubbo_JStack.log" + "." + dateStr));
-                    JVMUtil.jstack(jstackStream);
-                } catch (Throwable t) {
-                    logger.error("dump jstack error", t);
-                } finally {
-                    guard.release();
-                    if (jstackStream != null) {
-                        try {
-                            jstackStream.flush();
-                            jstackStream.close();
-                        } catch (IOException e) {
-                        }
-                    }
-                }
-
-                lastPrintTime = System.currentTimeMillis();
+        ExecutorService pool = Executors.newSingleThreadExecutor();
+        pool.execute(() -> {
+            String dumpPath = url.getParameter(Constants.DUMP_DIRECTORY, System.getProperty("user.home"));
+
+            SimpleDateFormat sdf;
+
+            String os = System.getProperty("os.name").toLowerCase();
+
+            // window system don't support ":" in file name
+            if (os.contains("win")) {
+                sdf = new SimpleDateFormat("yyyy-MM-dd_HH-mm-ss");
+            } else {
+                sdf = new SimpleDateFormat("yyyy-MM-dd_HH:mm:ss");
+            }
+
+            String dateStr = sdf.format(new Date());
+            //try-with-resources
+            try (FileOutputStream jStackStream = new FileOutputStream(new File(dumpPath, "Dubbo_JStack.log" + "." + dateStr))) {
+                JVMUtil.jstack(jStackStream);
+            } catch (Throwable t) {
+                logger.error("dump jStack error", t);
+            } finally {
+                guard.release();
             }
+            lastPrintTime = System.currentTimeMillis();
         });
+        //must shutdown thread pool ,if not will lead to OOM
+        pool.shutdown();

```

Comment:
How do we determine that the dump work has been completed?
The best practice should be to close the thread pool after the dump ends.

Reply:
Use the shutdown method instead of the shutdownnow method. Tasks previously submitted are executed, even in queues